File: app/src/main/java/com/example/flashcardapp/main/AddFlashcardActivity.java
// File: AddFlashcardActivity.java
package com.example.flashcardapp.main;

import android.os.Bundle;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import com.example.flashcardapp.R;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;

import com.example.flashcardapp.viewmodel.AddFlashcardViewModel;

public class AddFlashcardActivity extends AppCompatActivity {

    private EditText etQuestion, etAnswer;
    private Button btnSave;

    // Removed direct FlashcardDAO field
    // private FlashcardDAO flashcardDAO;

    // New: reference our ViewModel
    private AddFlashcardViewModel addFlashcardViewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_add_flashcard);

        etQuestion = findViewById(R.id.et_question);
        etAnswer = findViewById(R.id.et_answer);
        btnSave = findViewById(R.id.btn_save);

        // Removed old DAO init:
        // flashcardDAO = new FlashcardDAO(this);
        // flashcardDAO.open();

        // Now we set up the ViewModel
        addFlashcardViewModel = new ViewModelProvider(this).get(AddFlashcardViewModel.class);
        addFlashcardViewModel.initialize(new FlashcardDAO(this));

        // Observe saveComplete to react when insert is finished
        addFlashcardViewModel.getSaveComplete().observe(this, isComplete -> {
            if (isComplete != null && isComplete) {
                Toast.makeText(this, "Flashcard saved!", Toast.LENGTH_SHORT).show();
                finish();
            }
        });

        btnSave.setOnClickListener(v -> {
            String question = etQuestion.getText().toString().trim();
            String answer = etAnswer.getText().toString().trim();

            if (!question.isEmpty() && !answer.isEmpty()) {
                // Delegate to ViewModel
                addFlashcardViewModel.saveFlashcard(question, answer);
            } else {
                Toast.makeText(this, "Please enter both question and answer.", Toast.LENGTH_SHORT).show();
            }
        });
    }

    @Override
    protected void onDestroy() {
        // Removed old close call:
        // flashcardDAO.close();
        super.onDestroy();
    }
}

File: app/src/main/java/com/example/flashcardapp/main/AddFlashcardActivity.kt
package com.example.flashcardapp.main

import android.os.Bundle
import android.widget.Button
import android.widget.EditText
import android.widget.Toast

import androidx.lifecycle.ViewModelProvider
import com.example.flashcardapp.R
import com.example.flashcardapp.data.FlashcardRepository
import com.example.flashcardapp.data.FlashcardRoomDatabase
import com.example.flashcardapp.viewmodel.AddFlashcardViewModel
import com.example.flashcardapp.viewmodel.AddFlashcardViewModelFactory

class AddFlashcardActivity : BaseActivity() {

    private lateinit var etQuestion: EditText
    private lateinit var etAnswer: EditText
    private lateinit var btnSave: Button

    private lateinit var addFlashcardViewModel: AddFlashcardViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_add_flashcard)

        etQuestion = findViewById(R.id.et_question)
        etAnswer = findViewById(R.id.et_answer)
        btnSave = findViewById(R.id.btn_save)

        // Build repository
        val repository = FlashcardRepository(
            FlashcardRoomDatabase.getDatabase(applicationContext).flashcardDao()
        )
        // Create ViewModel using a custom Factory
        addFlashcardViewModel = ViewModelProvider(
            this,
            AddFlashcardViewModelFactory(repository)
        )[AddFlashcardViewModel::class.java]

        // Observe when saving is complete
        addFlashcardViewModel.saveComplete.observe(this) { isComplete ->
            if (isComplete == true) {
                Toast.makeText(this, "Flashcard saved!", Toast.LENGTH_SHORT).show()
                finish()
            }
        }

        btnSave.setOnClickListener {
            val question = etQuestion.text.toString().trim()
            val answer = etAnswer.text.toString().trim()
            if (question.isNotEmpty() && answer.isNotEmpty()) {
                addFlashcardViewModel.saveFlashcard(question, answer)
            } else {
                Toast.makeText(
                    this,
                    "Please enter both question and answer.",
                    Toast.LENGTH_SHORT
                ).show()
            }
        }
    }
}
